torch_dicom/inference/pipeline.py

```diff
@@ -0,0 +1,320 @@
+import logging
+from abc import ABC, abstractmethod
+from copy import copy
+from dataclasses import dataclass, field
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, Iterable, Iterator, List, Optional, Sequence, Tuple, TypeVar, Union
+
+import torch
+import torch.nn as nn
+from dicom_utils.dicom import Dicom
+from dicom_utils.volume import ReduceVolume, VolumeHandler
+from PIL import Image
+from torch import Tensor
+from torch.utils.data import ChainDataset, ConcatDataset, DataLoader, Dataset
+from torchvision.transforms.v2 import Compose
+from tqdm import tqdm
+
+from torch_dicom.datasets import (
+    DicomInput,
+    DicomPathDataset,
+    DicomPathInput,
+    ImageInput,
+    ImagePathDataset,
+    ImagePathInput,
+    TensorInput,
+    TensorPathDataset,
+    TensorPathInput,
+    collate_fn,
+    uncollate,
+)
+from torch_dicom.datasets.dicom import Transform
+from torch_dicom.preprocessing import MinMaxCrop, Resize
+
+
+B = TypeVar("B")
+
+
+@dataclass(repr=False)
+class InferencePipeline(ABC):
+    """
+    Inference pipeline for processing medical images.
+    It supports multiple input sources including DICOM files, tensors, and images.
+    The pipeline is designed to be used with PyTorch models and can handle batch processing.
+    It also supports custom datasets and transformations. It is assumed that models will return predictions as a dictionary.
+
+    The pipeline is an iterator that yields input examples and predictions for each example.
+
+    Args:
+        dicom_paths: Paths to DICOM files to be processed.
+        dicoms: DICOM objects to be processed.
+        tensor_paths: Paths to tensor files to be processed.
+        tensors: Tensor objects to be processed.
+        image_paths: Paths to image files to be processed.
+        images: PIL images to be processed.
+        custom_inputs: Custom datasets to be processed.
+        device: Device to be used for processing.
+        batch_size: Size of the batch for processing.
+        transform: Optional transformation to be applied to the inputs.
+        enumerate_inputs: Whether to enumerate the inputs.
+            This enables progress bars to display a total and ETA but may be slower to initialize.
+            Note that this will only work if all inputs are map-style datasets. When ``True``, ``skip_errors``
+            will not be applied to any enumerated datasets.
+        skip_errors: Whether to skip errors during processing.
+        volume_handler: Handler for volume data.
+        dataloader_kwargs: Additional arguments for the DataLoader.
+    """
+
+    # Known input sources
+    dicom_paths: Iterable[Path] = field(default_factory=list)
```

Comment:
This is fine, but it may be better to consider a simpler/easier input interface than maintain support for all of these input types in the future.

Reply:
Agreed. I don't see an immediate need to handle inputs other than DICOMs, tensors, and images. If this changes and the number of supported inputs continues to grow we should revisit input handling.